### PR TITLE
Update input superglobal access to use new Input class in all places.

### PR DESCRIPTION
**Pull request by aaemnnosttv**
## Summary

This PR picks up where #916 leaves off to update all remaining usage to use the new Input class via `Context`.

This PR currently uses `enhance/905-credentials-provisioning` as the base as that PR has not been merged yet so that only the added changes are shown here.

Addresses issue #886

## Relevant technical choices

* Updates all references to `$_GET`, and `$_SERVER` (nothing else to update)
* Updates all existing code using `filter_input`
* Also updates "manual" redirects using `header( 'Location: '` code to use `wp_safe_redirect` instead for consistency and testability

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 4.7 and PHP 5.4.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

File: includes/Core/Authentication/Authentication.php [hunk 1]
```diff
@@ -408,12 +408,12 @@ private function handle_oauth() {
 				)
 			);
 
-			header( 'Location: ' . filter_var( $redirect_url, FILTER_SANITIZE_URL ) );
```
Comment (felixarntz) [inline includes/Core/Authentication/Authentication.php]: Why remove this `filter_var()` call? That seems unrelated.

File: includes/Core/Authentication/Authentication.php [hunk 2]
```diff
@@ -422,13 +422,13 @@ private function handle_oauth() {
...
-			if ( ! empty( $_GET['redirect'] ) ) { // phpcs:ignore WordPress.Security.NonceVerification.NoNonceVerification
-				$redirect_url = esc_url_raw( wp_unslash( $_GET['redirect'] ) ); // phpcs:ignore WordPress.Security.NonceVerification.NoNonceVerification
+			$redirect_url = $input->filter( INPUT_GET, 'redirect' );
+			if ( $redirect_url ) {
+				$redirect_url = esc_url_raw( wp_unslash( $redirect_url ) );
 			}
 
 			// User is trying to authenticate, but access token hasn't been set.
-			header( 'Location: ' . filter_var( $auth_client->get_authentication_url( $redirect_url ), FILTER_SANITIZE_URL ) );
+			wp_safe_redirect( $auth_client->get_authentication_url( $redirect_url ) );
```
Comment (felixarntz) [inline includes/Core/Authentication/Authentication.php]: Same here.

File: includes/Core/Authentication/Authentication.php (hunk 1 above)
Comment (aaemnnosttv) [reply includes/Core/Authentication/Authentication.php]: In this case it's not necessary since the URL is from an internal source (`$this->context->admin_url()`). Also, WordPress core does not do this (which I found interesting) but I think it may be due to `wp_safe_redirect` calling `wp_sanitize_redirect` on the `$location` internally.

If we want to make it a policy to escape URLs in all redirects then I think we should standardize on using `esc_url_raw` for consistency.

File: includes/Core/Authentication/Authentication.php (hunk 2 above)
Comment (aaemnnosttv) [reply includes/Core/Authentication/Authentication.php]: I've updated this to be escaped now since the basis of the redirect URL can come from the request 👍

File: includes/Core/Authentication/Authentication.php [hunk 3]
```diff
@@ -525,11 +529,9 @@ private function inline_js_setup_data( $data ) {
 			$data['hasSearchConsoleProperty'] = false;
 		}
 
-		$reauth                        = isset( $_GET['reAuth'] ) ? ( 'true' === $_GET['reAuth'] ) : false; // phpcs:ignore WordPress.CSRF.NoNonceVerification.
-		$data['showModuleSetupWizard'] = $reauth;
+		$data['showModuleSetupWizard'] = $this->context->input()->filter( INPUT_GET, 'reAuth', FILTER_VALIDATE_BOOLEAN );
 
-		$module_to_setup       = isset( $_GET['slug'] ) ? sanitize_key( $_GET['slug'] ) : ''; // phpcs:ignore WordPress.CSRF.NoNonceVerification.
-		$data['moduleToSetup'] = $module_to_setup;
+		$data['moduleToSetup'] = sanitize_key( $this->context->input()->filter( INPUT_GET, 'slug' ) );
```
Comment (felixarntz) [inline includes/Core/Authentication/Authentication.php]: If `slug` is not set, this would cause a problem due to `sanitize_key()` not receiving a string. We need to make sure the value is a string, an empty string would be fine.

File: includes/Modules/Site_Verification.php [hunk 4]
```diff
@@ -75,10 +75,13 @@ function() {
 		add_action(
 			'init',
 			function () {
+				$request_uri    = $this->context->input()->filter( INPUT_SERVER, 'REQUEST_URI' );
+				$request_method = $this->context->input()->filter( INPUT_SERVER, 'REQUEST_METHOD' );
```
Comment (felixarntz) [inline includes/Modules/Site_Verification.php]: There are some known issues regarding `INPUT_SERVER` and `INPUT_ENV` in certain setups, I ran into this in another project. See e.g. https://stackoverflow.com/questions/25232975/php-filter-inputinput-server-request-method-returns-null for some information

We could either keep the `$_SERVER` accesses untouched for now, or add internal logic to the `Input` class to work around this issue. What do you think?

Comment (aaemnnosttv) [reply includes/Modules/Site_Verification.php]: Good catch! I wasn't aware of this. I would prefer to use the same utility in all cases. This wouldn't impact `MutableInput` so we only need to patch `Input` to ensure this works. I'll see what it would take to add 👍 

I think worst case, we could attempt to mimic the immutable behavior of `filter_input` by taking a copy of those particular superglobals when the class is instantiated. Then we could use `filter_var` internally like we do with `MutableInput`. Ideally we could detect when this would be necessary, if not, we can either try `filter_input` first with a fallback or just use the alternate method; with these input types only of course.

File: includes/Core/Authentication/Authentication.php (hunk 3 above)
Comment (aaemnnosttv) [reply includes/Core/Authentication/Authentication.php]: Updated!

**Comment by aaemnnosttv** [comment]
@felixarntz I've updated this with handling for environments where `filter_input` does not work properly with `$_SERVER` or `$_ENV`.

I did some testing around this and indeed it seems to be specific to environments using Fast CGI (my local was affected by this as well), although I didn't want to couple the fallback to that specifically. I found that the default behavior worked fine with Apache, or PHP built-in server (although hopefully no one's using that 😆).

The fallback is thoroughly documented so you can see exactly what's happening there. Like I mentioned before, we'll try to mimic the immutable behavior of `filter_input` by taking a copy of these superglobals when instantiated. It's not 100% as good as `filter_input`, but it's the best we can do within reason 😄

You can test this locally like so

```php
add_action( 'googlesitekit_init', function () {
	$input = new \Google\Site_Kit\Core\Util\Input();

	if ( $key = $input->filter( INPUT_GET, 'key' ) ) {
		var_dump( $input->filter( INPUT_SERVER, $key ) );
	}

	var_dump( $_SERVER );
	exit;
} );
```

Should be ready for another look!

File: includes/Core/Util/Input.php [hunk 5]
```diff
@@ -18,6 +18,26 @@
  * @ignore
  */
 class Input {
+	/**
+	 * Map of input type to superglobal array.
+	 *
+	 * For use as fallback only.
+	 *
+	 * @var array
```
Comment (felixarntz) [inline includes/Core/Util/Input.php]: Missing `since` :)

File: includes/Core/Util/Input.php [hunk 6]
```diff
@@ -39,6 +59,20 @@ class Input {
 	 *                                      and NULL if the filter fails.
 	 */
 	public function filter( $type, $variable_name, $filter = FILTER_DEFAULT, $options = null ) {
-		return filter_input( $type, $variable_name, $filter, $options );
+		$value = filter_input( $type, $variable_name, $filter, $options );
+
+		// Fallback for environments where filter_input may not work with SERVER or ENV types.
+		if (
+			// Only use this fallback for affected input types.
+			in_array( $type, array( INPUT_SERVER, INPUT_ENV ), true )
```
Comment (felixarntz) [inline includes/Core/Util/Input.php]: I think we can strip this check and only rely on the `array_key_exists` check below which results in the same.

File: includes/Core/Util/Input.php [hunk 7]
```diff
@@ -18,6 +18,26 @@
...
+	 */
+	protected $map;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 */
+	public function __construct() {
+		$this->map = array(
```
Comment (felixarntz) [inline includes/Core/Util/Input.php]: Maybe we should add a comment above this hinting at the bug with `INPUT_ENV` and `INPUT_SERVER` (to explain why only those are here).
